### PR TITLE
HWKAPM-776 Don't record 'null' valued service and buildStamp properties

### PR DESCRIPTION
**Pull request by objectiser**

### CODE REVIEW

**Comment by objectiser** [comment]
@pavolloffay Could you review please?

Thought best to keep a Property per Tag for now, regardless of whether the value happens to be null - i.e. leave that for the opentracing spec/abstract impl to control. So just put a guard around the service and buildStamp instead.

File: client/opentracing/src/main/java/org/hawkular/apm/client/opentracing/EnvironmentAwareTraceRecorder.java [hunk 1]
```diff
@@ -38,10 +38,12 @@ public void record(Trace trace) {
         if (trace.getNodes() != null && trace.getNodes().size() > 0) {
             Node rootNode = trace.getNodes().get(0);
             if (this.deploymentMetaData != null) {
-                if (!rootNode.hasProperty(Constants.PROP_SERVICE_NAME)) {
+                if (!rootNode.hasProperty(Constants.PROP_SERVICE_NAME)
```
Comment (pavolloffay) [inline client/opentracing/src/main/java/org/hawkular/apm/client/opentracing/EnvironmentAwareTraceRecorder.java]: It is more efficient to first test `null != this.deploymentMetaData.getServiceName()` then iterate over the props. Same on line `45`.

File: client/opentracing/src/test/java/org/hawkular/apm/client/opentracing/EnvironmentAwareTraceRecorderTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                new Property(Constants.PROP_BUILD_STAMP, "myBuildStamp"))),
+                getTraceProperties(new DeploymentMetaData("myService", "myBuildStamp")));
+    }
+
+    @Test
+    public void testDeploymentMetaDataNotSet() {
+        assertTrue(getTraceProperties(new DeploymentMetaData(null, null)).isEmpty());
+    }
+
+    protected Set<Property> getTraceProperties(DeploymentMetaData dmd) {
```
Comment (pavolloffay) [inline client/opentracing/src/test/java/org/hawkular/apm/client/opentracing/EnvironmentAwareTraceRecorderTest.java]: Wouldn't be better to call it `getDeploymentMedaData` properties?

File: client/opentracing/src/main/java/org/hawkular/apm/client/opentracing/EnvironmentAwareTraceRecorder.java (hunk 1 above)
Comment (objectiser) [inline client/opentracing/src/main/java/org/hawkular/apm/client/opentracing/EnvironmentAwareTraceRecorder.java]: The service name may be provided by the application (i.e. setting tag directly).

File: client/opentracing/src/test/java/org/hawkular/apm/client/opentracing/EnvironmentAwareTraceRecorderTest.java (hunk 2 above)
Comment (objectiser) [inline client/opentracing/src/test/java/org/hawkular/apm/client/opentracing/EnvironmentAwareTraceRecorderTest.java]: It is not returning the deployment metadata - although trace properties is misleading aswell. It is actually returning the properties associated with the trace root node, so `getTraceRootNodeProperties`?

File: client/opentracing/src/main/java/org/hawkular/apm/client/opentracing/EnvironmentAwareTraceRecorder.java (hunk 1 above)
Comment (pavolloffay) [inline client/opentracing/src/main/java/org/hawkular/apm/client/opentracing/EnvironmentAwareTraceRecorder.java]: Sure, but `if (!rootNode.hasProperty(Constants.PROP_SERVICE_NAME)  && this.deploymentMetaData.getServiceName() != null)`

is the same as:

`if( this.deploymentMetaData.getServiceName() != null && !rootNode.hasProperty(Constants.PROP_SERVICE_NAME) )`

but later does not iterate(see `node#hasProperty()`) over properties if serviceName in deploymentMetaData is null.

File: client/opentracing/src/test/java/org/hawkular/apm/client/opentracing/EnvironmentAwareTraceRecorderTest.java (hunk 2 above)
Comment (pavolloffay) [inline client/opentracing/src/test/java/org/hawkular/apm/client/opentracing/EnvironmentAwareTraceRecorderTest.java]: Actually some similar tests are in `DeploymentMedaDataTest`, the last two test test `EnvironmentAwareTraceRecorder` do we want to move those tests to this class? @jpkrohling ?

Comment (objectiser) [inline client/opentracing/src/test/java/org/hawkular/apm/client/opentracing/EnvironmentAwareTraceRecorderTest.java]: I think these are fine where they are, as they are focused more on the deployment meta data provided to the tracer.

**Comment by objectiser** [comment]
retest this please

**Comment by objectiser** [comment]
@pavolloffay Thanks for the review.

@jpkrohling If you think the tests pointed out by Pavol should be relocated, then we can do it in a separate PR.
